Remove Armadillo

michaelweylandt opened the pull request:
This PR removes all `arma` code and uses `Eigen` only. In general, it seems to significantly speed things up, but I haven't done formal benchmarking. 

I'm encountering a bug I can't figure out: 

`R> CARP(presidential_speech, alg.type="carp")`

doesn't work. (It does, however, work if a set a small value of `t`). Any thoughts? I don't think this is a function of my changes, but not sure....

jjn13 commented on src/carp.cpp:
```diff
@@ -0,0 +1,163 @@
+  }
+
+  // Now that we are done, we can "drop" unused buffer space before returning to R
+  UPath.conservativeResize(UPath.rows(), path_iter - 1);
```
I think the resize here cuts off the last iteration in the path. I tested locally by changing to:

UPath.conservativeReize(UPath.rows(),path_iter);
VPath.conservativeReize(VPath.rows(),path_iter);
gamma_path.conservativeReize(gamma_Path.rows(),path_iter);
vZerodInds_Path.conservativeReize(vZeroInds_Path.rows(),path_iter);

and the CARP tests pass afterwards.

jjn13 replied:
I will try to see if this work for CBASS too.

michaelweylandt replied:
A gentleman and a scholar!

As confirmed by adding the lines

```
if(UPath.col(UPath.cols()) != u_new){
    Rcpp::stop("Oh SNAP!"); 
}
```

reveals, I was indeed dropping the last iteration. 

This is a classic off-by-one error: `path_iter` is the zero-based index of the next column we want to insert to, while `conservativeResize` takes the (one-based) number of columns we want to keep. 

If, for example, our algorithm terminated after 100 save-worthy iterations, we would have `path_iter = 100` (because the possible next iteration would be in column 100, *i.e.*, `UPath.col(101)`), so `path_iter - 1 = 99` and we are loosing a column.

michaelweylandt replied:
Addressed in 116227e

michaelweylandt commented:
Fixed a few more mistakes: as of now, the CARP variants pass the `all.equal` test vs the current version. (Tested on `presidential_speech` for `CARP-VIZ` as well as `CARP` at `t=1.2, 1.1, 1.05`.)

The CBASS variants are returning different results because I changed the behavior to keep every `keep` iterations for consistency with the CARP functions. Hence, we get different path lengths and I can't really compare them. Any thoughts on how to test those?

michaelweylandt commented:
I just removed the code to store results every `keep` iterations and now both the `CBASS` and `CARP` variants match the current Armadillo code (up to `all.equal` standards). 

In particular, the result of 

```
list(
  CARPVIZ   = CARP(presidential_speech, alg.type="carp.viz")$carp.sol.path), 
  CARP1.2   = CARP(presidential_speech, alg.type="carp", t=1.2)$carp.sol.path), 
  CARP1.1   = CARP(presidential_speech, alg.type="carp", t=1.1)$carp.sol.path), 
  CARP1.05  = CARP(presidential_speech, alg.type="carp", t=1.05)$carp.sol.path), 
  CBASSVIZ  = CBASS(presidential_speech, alg.type="cbass.viz")$cbass.sol.path), 
  CBASS1.2  = CBASS(presidential_speech, alg.type="cbass", t=1.2)$cbass.sol.path), 
  CBASS1.1  = CBASS(presidential_speech, alg.type="cbass", t=1.1)$cbass.sol.path), 
  CBASS1.05 = CBASS(presidential_speech, alg.type="cbass", t=1.05)$cbass.sol.path)
)
```

is the same (up to `all.equal` standards) before and after this PR. 

It takes 37s now, compared to 77s before, so that's a nice bump!

michaelweylandt commented:
Any more review comments?

On Fri, Sep 7, 2018 at 08:22 jjn13 <notifications@github.com> wrote:

> Wowsers that's quite the speed-up 😀
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jjn13/clustRviz/pull/6#issuecomment-419437311>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABau6UFgaMu9-6fY5EIfay35NN02Yf48ks5uYnL4gaJpZM4WVl9t>
> .
>
